File: leetcode/sort-list/sort-list.cpp
/**************************************
* @author dujiong
* @date 2016.9.20
* @version V0.1
**************************************/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *sortList(ListNode *head) {
        if(!head || !head->next) return head;
        ListNode* p=head;
        ListNode* q=head->next;
        
        while(q && q->next){	//快慢指针思路，慢指针来找到链表的中间节点
            p = p->next;
            q = q->next->next;
        }
        ListNode* right = sortList(p->next);
        p->next=NULL;
        ListNode* left = sortList(head);
        return merge(left, right);
     }
 	 
private:
    ListNode* merge(ListNode* left, ListNode* right){
        ListNode dummy(0);
        ListNode* p = &dummy;
        while(left && right){
            if(left->val < right->val){
                p->next = left;
                left = left->next;
            }else{
                p->next = right;
                right = right->next;
            }
            p = p->next;
        }
        if(left) p->next = left;
        if(right) p->next = right;
        return dummy.next;
    }
       
};

File: leetcode-golang/Sort List/Sort List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func sortList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }
    //fast = head.Next
    slow, fast := head, head.Next
    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
    }
    //slow是前半链表的最后一个节点
    right := sortList(slow.Next)
    slow.Next = nil
    left := sortList(head)
    return merge(left, right)
}

func merge(left *ListNode, right *ListNode) *ListNode {
    dummy := new(ListNode)
    cur := dummy
    for left != nil && right != nil {
        if left.Val <= right.Val {
            cur.Next = left
            left = left.Next
        } else {
            cur.Next = right
            right = right.Next
        }
        cur = cur.Next
    }
    if left != nil {
        cur.Next = left
    } else if right != nil {
        cur.Next = right
    }
    return dummy.Next
}